src/cli/serve/serve.js

```diff
@@ -128,45 +133,96 @@ module.exports = function (program) {
       }
     }
 
-    const getCurrentSettings = () => readServerSettings(opts, this.getUnknownOptions());
-    const settings = getCurrentSettings();
+    if (opts.dev && opts.multiConfigDir) {
+      logFatal('--multi-config-dir can not be used in dev mode');
+      process.exit(1);
+    }
+
+    const getSettings = () => readServerSettings(opts, this.getUnknownOptions());
 
     if (canCluster && opts.dev && !isWorker) {
       // stop processing the action and handoff to cluster manager
       const ClusterManager = require('../cluster/cluster_manager');
-      new ClusterManager(opts, settings);
+      new ClusterManager(opts, getSettings());
       return;
     }
 
-    let kbnServer = {};
-    const KbnServer = require('../../server/kbn_server');
-    try {
-      kbnServer = new KbnServer(settings);
-      await kbnServer.ready();
+    let reloadServerSettingsFns;
+    if (!opts.multiConfigDir) {
+      reloadServerSettingsFns = await startServers([getSettings]);
+    } else {
+      reloadServerSettingsFns = await startServers(
+        readdirSync(opts.multiConfigDir)
+        .map(name => {
+          if (name.startsWith('.')) return null;
+
+          const multiConfigPath = resolve(opts.multiConfigDir, name);
+          const getMultiSettings = () => {
+            const settings = getSettings();
+            const multiOpts = {
+              ...opts,
+              config: multiConfigPath
+            };
+            const multiSettings = readServerSettings(multiOpts, {});
+            return _.defaultsDeep(multiSettings, settings);
+          };
+
+          return getMultiSettings;
+        })
+        .filter(Boolean)
+      );
     }
-    catch (err) {
-      const { server } = kbnServer;
 
-      if (err.code === 'EADDRINUSE') {
-        logFatal(`Port ${err.port} is already in use. Another instance of Kibana may be running!`, server);
-      } else {
-        logFatal(err, server);
-      }
+    process.on('SIGHUP', function reloadConfig() {
```

Comment:
this is just a small thing: I don't think windows has a SIGHUP signal. I believe node simulates it when closing the terminal, but we probably also want it on termination of the process (E.g. ctrl-c in cygwin). Not sure what all the ins&outs are, but can look into that if you want.